torch/autograd/_functions/pointwise.py

```diff
@@ -138,103 +156,123 @@ def backward(ctx, grad_output):
         return i.cos()
 
-    def backward(self, grad_output):
-        i, = self.saved_tensors
+    @staticmethod
+    def backward(ctx, grad_output):
+        i, = ctx.saved_variables
         return grad_output.mul(i.sin()).neg_()
 
 
 class Tan(Function):
 
-    def forward(self, i):
-        self.save_for_backward(i)
+    @staticmethod
+    def forward(ctx, i):
+        ctx.save_for_backward(i)
         return i.tan()
 
-    def backward(self, grad_output):
-        i, = self.saved_tensors
+    @staticmethod
+    def backward(ctx, grad_output):
+        i, = ctx.saved_variables
         return grad_output.div(i.cos().pow(2))
 
 
 class Asin(Function):
 
-    def forward(self, i):
-        self.save_for_backward(i)
+    @staticmethod
+    def forward(ctx, i):
+        ctx.save_for_backward(i)
         return i.asin()
 
-    def backward(self, grad_output):
-        i, = self.saved_tensors
-        return grad_output * (1 - i.mul(i)).sqrt_().reciprocal_()
+    @staticmethod
+    def backward(ctx, grad_output):
+        i, = ctx.saved_variables
+        return grad_output * (1 - i.mul(i)).sqrt().reciprocal()
 
 
 class Acos(Function):
 
-    def forward(self, i):
-        self.save_for_backward(i)
+    @staticmethod
+    def forward(ctx, i):
+        ctx.save_for_backward(i)
         return i.acos()
 
-    def backward(self, grad_output):
-        i, = self.saved_tensors
-        return grad_output.mul((1 - i.mul(i)).sqrt_().reciprocal_()).neg_()
+    @staticmethod
+    def backward(ctx, grad_output):
+        i, = ctx.saved_variables
+        return grad_output.mul((1 - i.mul(i)).sqrt().reciprocal()).neg()
 
 
 class Atan(Function):
 
-    def forward(self, i):
-        self.save_for_backward(i)
+    @staticmethod
+    def forward(ctx, i):
+        ctx.save_for_backward(i)
         return i.atan()
 
-    def backward(self, grad_output):
-        i, = self.saved_tensors
-        return grad_output * i.mul(i).add_(1).reciprocal_()
+    @staticmethod
+    def backward(ctx, grad_output):
+        i, = ctx.saved_variables
+        return grad_output * i.mul(i).add(1).reciprocal()
 
 
 class Reciprocal(Function):
 
-    def forward(self, i):
+    @staticmethod
+    def forward(ctx, i):
         result = i.reciprocal()
-        self.save_for_backward(result)
+        ctx.save_for_backward(result)
         return result
 
-    def backward(self, grad_output):
-        result, = self.saved_tensors
+    @staticmethod
+    def backward(ctx, grad_output):
+        result, = ctx.saved_variables
         return grad_output * result.mul(result).neg_()
 
 
 class Cmax(Function):
 
-    def forward(self, a, b):
-        self._max_buffer = a.gt(b).type_as(a)
+    @staticmethod
+    def forward(ctx, a, b):
+        ctx.save_for_backward(a, b)
         return a.max(b)
 
-    def backward(self, grad_output):
+    @staticmethod
+    def backward(ctx, grad_output):
+        a, b = ctx.saved_variables
+        mask = a.gt(b).type_as(a)
```

Comment:
It's better to save the mask in forward (just `a.gt(b)` and convert it to `grad_output`'s type here (it's going to save a lot of memory - if a and b are float tensors it's an 8x reduction!). To increase the efficiency even more you can guard the `ctx.mask = a.gt(b)` with an `if ctx.requires_grad`